### ISSUE TITLE
Core DOM class names to match WebKit?

### ISSUE THREAD

**Issue by cpboyd** [SVGKit/SVGKit#442]
I noticed this after using another framework that also had a "Text" class.

Is there any specific reason why the class names don't match WebKit framework class names (e.g. DOMText)?

Perhaps utilizes the actual WebKit classes for the OS X builds.

**Comment by adamgit** [comment]
DOMText is the wrong name. Webkit uses that to workaround major bugs in
ObjectiveC (ie. the total lack of namespacing), but it's unofficial.

I've tried a few times to alias class names at rutnime, so that you could
dynamically replace which DOM implementation you're using, but I can't get
it to work - Xcode crashes for me with numerous bugs and intenral Apple
errors. Maybe I was doing it wrong, but no-one else has stepped up to try!

For what it's worth, I think it would be great to use WebKit's classes
where available - but how can we do this without rewriting tonnes of code?

**Comment by cpboyd** [comment]
I can look into the aliasing, because I think it'd be worthwhile for the OS X version to use WebKit.
(And, importantly, make it easier to "upgrade" the iOS version if Apple ever decides to add the DOM classes to the iOS WebKit framework.)

But I still think it's worth considering refactoring the existing class names for the same reason that Apple did: Lack of namespacing.

A few Java libraries (which does have namespacing) even use "DOM<NodeName>" class names for XML parsing, and PHP (which also supports namespacing) has a DOMText.

Just "Text" seems simple enough that there's bound to be conflicts with other libraries.

**Comment by MiMo42** [comment]
@cpboyd +1

**Comment by adamgit** [comment]
But ... DOMText will cause problems too, for same reason. Leaving us with
... "SVGKITDOM" as a prefix? Hellish :(.

**Comment by MiMo42** [comment]
"Don't fight the frameworks" :-)
I'd really prefer to follow Apple's naming conventions on DOM classes.
We should use them where available (OS X) and add a compatibility layer of the missing classes as long as they are private (iOS). I did this with other classes before. It is fairly easy to replace the  compatibility layer as soon as the classes become public.
I also would like to extract the compatibility layer as a whole, so others can opt out and use their own.

**Comment by cpboyd** [comment]
I don't think we need to go with "SVGKITDOM"...

I think "DOM" is fine, as it should drastically limit the potential conflicts.

From a quick Google search, I could only find one library from 2005 that used those class names:
https://github.com/inverse-inc/sope/tree/master/sope-xml/DOM

Moreover, there shouldn't be any worry about "DOM" class ambiguity on iOS as long as we mimic the official WebKit DOM classes.  (Since anyone else using the "DOM" prefix is likely doing it for the same reason that we are.)

I'll work on this in a fork.

**Comment by MaddTheSane** [comment]
I think just prefixing classes with SVGK would be enough.

**Comment by cpboyd** [comment]
@MaddTheSane I agree, if the desire is for the classes to remain distinct from WebKit's classes.

"SVGK" (without "DOM") would certainly be unique enough.

**Comment by adamgit** [comment]
I just want to get rid of the DOM implementation entirely - or split it off
as a separate github project, and ship a binary / Framework of it with
SVGkit, in a way that SVGKit users can opt to completely replace it with
ease.

I really don't want us to be taking responsibility for maintaining a full
DOM implementation! What we have I wrote 99% myself over a weekend or two,
simply as a shim to get us closer to SVG Spec compliance. It could be
waaaaay better (I hate to think what it would do on a large SVG which uses
CSS heavily - massive mem usage? Very slow parse speeds? I was churning out
code as fast as I could, with only one rule: "does it adhere to the
official DOM spec? If so: it's in")

**Comment by MiMo42** [comment]
@cpboyd  
For the records: https://github.com/inverse-inc/sope/tree/master/sope-xml/DOM
is using protocols that classes conform to.
Though file name reads DOMDocument.x, the code reads:
@interface NGDOMDocument : NGDOMNodeWithChildren < DOMDocument >
/_..._/
@end
They use NGDOM as a prefix for classes that conform to a protocol that actually describes the DOM interface.
The DOMxxx protocols correspond to a DOM interface and the NGDOMDocument acts as an implementation that provides one or more DOM interfaces. 
Looks cool at first. The downside is: Objective-C does not allow default implementations for methods defined in a protocol (like in swift).

**Comment by cpboyd** [comment]
@adamgit
I also think that it makes a lot of sense to split the DOM implementation off into its own Github project and Framework.

From that perspective, I'd still lean towards having it be like a public extension to the WebKit library for iOS.
(Certainly other projects likely could benefit from the DOM parsing functionality as well.)

**Comment by MiMo42** [comment]
@adamgit 
On memory consumption and speed: 
NSXMLParser supports Event-Driven XML Parsing. It is memory efficient and fast.
Maybe this is an option to check out later on.

**Comment by cpboyd** [comment]
@MiMo42 
Honestly, I didn't really look at the files themselves except to see that the majority of copyright statements are from 2005.

It looks like there's still activity on-going (it has a commit from an hour ago) in the library, but it's an OS X library that would conflict with the official WebKit.

My opinion is that SVGKit should split off @adamgit's DOM stuff as a separate project/framework that's compatible with the official WebKit framework.

That would ease the transition to WebKit should Apple make the classes available on iOS (and allow someone compiling for OS X to switch to the official OS X WebKit.framework.)

Alternatively, it might be possible to use NSXMLParser... But it would probably require significant re-work of code.

**Comment by adamgit** [comment]
Yep, so priority #1: can anyone successfully refactor our DOM
implementation so that:
1. The class names that SVGKit classes have to reference are sensible (e.g.
   "DOMDocument")
2. The actual classes can be named anything (e.g. "AdamGitDOMDocument" )
3. Replacing the classes (e.g. AdamGitDOMDocument) with alternative
   implementation (e.g. BoydDOMDocument) is very quick / easy for a developer
   who's importing the SVGKit framework to their project

3 above should _ideally_ be:
a. drag/drop a different framework (instead of "AdamGitDOM.framework") into
your project
b. BAM! ... xcode links the other framework, SVGKit doesn't know or care -
it carries on working fine

...this is what I could never quite get to work. I concluded that
ObjectiveC is a vile language for any medium/large scale projects until
someone adds namespacing (Apple? APPLE!?? Get on with it!), we should all
just give up :). So I put down the fork, walked slowly away from the
keyboard, and pretended it was all a bad dream.

**Comment by MiMo42** [comment]
There ist no runtime symbol mapping. The symbols are baked into the frameworks and have to be imported at compile time and have to match. 
We could do funny runtime stuff, but that road would end up in weird code that no one wants to maintain.
Keep it simple. We should just add an additionally framework that uses the symbols apple does.
Your existing code with an updated set of symbols.
If someone wants to opt in another self-baked framework, they have to add a preprocessor-based mapping at compile-time on their side (which is easy) ;-)

**Comment by cpboyd** [comment]
@adamgit I think I've seen where Swift is supposed to have namespacing for frameworks.
However, that doesn't really benefit SVGKit unless it's fully ported to Swift.

I've refactored the class names to match WebKit in my fork:
https://github.com/cpboyd/SVGKit

**Comment by adamgit** [comment]
@MiMo42 - fine, but we need instructions to be included in the Pull Request
that merges these changes (i.e. it has to change the README in the root of
the project) so that people know exactly how to use it.

**Comment by MiMo42** [comment]
Maybe you should consolidate branches to get master the master again ;-) 
You really should create a new release before we go on to make DOM stuff an external framework.
As you already have created a repository for it (DOMiOS): Are we gonna use that one?

**Comment by adamgit** [comment]
"opt out and use their own" is what I really want - it should be possible 
to plug and play different DOM implementations.

But so far I cannot find anyway of getting objc to link this. its depressing.

**Comment by MiMo42** [comment]
OK, I digged somewhat deeper and this is want I found:
As documented and due to simplifications, the SVGKit's Core DOM implementation is still too far away from being replaced with ease. 
There are non-standard extensions and even public initializers for DOM classes that should not have one (@adamgit: I found you already mentioned this in code ;-) 
So before we can replace the existing core DOM impl, we'd need to change the existing parts of the implementation to follow the w3c specs purely. That does not mean to create a full blown impl, but to ensure that SVGKit uses all DOM interfaces as intended and that the existing parts of SVGKit's Core DOM impl implement the standard without SVGKit specific extensions. 
All necessary extensions of Core DOM classes should be added as categories or subclasses to SVGKit, not to the (then extractable) Core DOM impl.

Here's what I've done with SVGKit today: 
1) I created a module enabled iOS Framework using the files prefixed by @cpboyd. 
2) I _tried_ to switch to WebKit.framework's implementation (on OS X, obviously).
3) Driving home now ;-)
